Writerside/topics/raptor-xx.md

```diff
@@ -1,3 +1,435 @@
+  filtering and transforming. With Ranges Views, transformations occur on the original range without creating new
+  containers or copies unless explicitly needed, thereby improving both memory and runtime efficiency.
+
+#### Example Usage of Ranges Views in Our Code
+
+In our implementation, Ranges Views allowed us to simplify complex operations like filtering trips, sorting schedules,
+or transforming datasets. For example:
+
+```c++
+const auto servedDates = data->calendarDates
+                             | std::views::values
+                             | std::views::join
+                             | std::views::filter([this, &localDateTime](const std::shared_ptr<schedule::gtfs::CalendarDate>& date) {
+                                 return this->isServiceAvailable(date->serviceId, localDateTime);
+                               })
+                             | std::views::transform([](const std::shared_ptr<schedule::gtfs::CalendarDate>& date) {
+                                 return date->serviceId;
+                               })
+                             | std::ranges::to<std::set<std::string>>();
+
+```
+
+### Profiling
+
+## Visual Studio Diagnostic Tools
+
+We used the Visual Studio Diagnostic Tool to analyze performance.
+The **Visual Studio Diagnostic Tools** are a powerful set of profiling and debugging features integrated into the Visual
+Studio IDE. These tools allow developers to monitor, analyze, and optimize the performance of their applications,
+including C++ code. By leveraging these tools, you can gain deeper insights into how your program behaves at runtime,
+helping you identify and resolve performance bottlenecks and resource management issues.
+
+### Key Features for C++ Code
+
+1. **Performance Profiling**:
+
+- The profiling tools provide detailed performance metrics such as CPU usage, memory allocation, and execution time of
+  different code sections. This helps in identifying functions or loops that consume excessive processing time or
+  memory.
+
+2. **Memory Usage Analysis**:
+
+- The diagnostic tools track memory allocations in the running C++ application. This is especially useful for detecting
+  memory leaks, inefficient memory usage, or unexpected allocations, ensuring the program remains efficient and stable.
+
+3. **CPU Usage**:
+
+- The CPU usage tool allows to see how much processing time the code consumes, and which parts of the code are
+  responsible for high CPU loads. This enables to focus optimization efforts on the most critical areas.
+
+- [Optimizing Code Using Profiling Tools](https://learn.microsoft.com/en-us/visualstudio/profiling/optimize-code-using-profiling-tools?view=vs-2022&source=recommendations).
+
+![cpp-diagnostic.png](cpp-diagnostic.png)(compiler_support.png){width="850"}
+![cpp-diagnostic-scanRoute.png](cpp-diagnostic-scanRoute.png){width="850"}
+
+## Foreign Function and Memory (FFM) API
```

Comment:
I think this section does not really explain why we didn't implement the FFM API. The reason was that sending routing requests from JAVA to C++, relied on C++ being able to operate on the schedule memory layout built in JAVA. Since JAVA would be the only active application, alternatively building the schedule from files in C++ from files for each request (not performant at all). The better solution would have been having a C++ service, which Java could address when needed for routing.

Summary, FFM API is not suited for the application we were developing. It's better to replace complex functions instead of whole packages.